File: DEMOAPPS/DEMOAPP001_TYPESCRIPT_CYPRESS/app_src/PuzzleLoader.ts
import * as fs from "fs";
import * as path from "path";
import { GRID_SIZE, EMPTY_CELL, MAX_DIGIT } from "./constants";

/**
 * Represents a single puzzle with metadata
 */
export interface Puzzle {
    name: string;
    difficulty: string;
    description: string;
    grid: number[][];
}

/**
 * Structure of the puzzles.json file
 */
interface PuzzleCollection {
    puzzles: Puzzle[];
}

/**
 * Utility class for loading Sudoku puzzles from JSON files.
 * Follows the Single Responsibility Principle: Data Loading Only.
 */
export class PuzzleLoader {
    private puzzles: Puzzle[] = [];

    /**
     * Loads puzzles from the specified JSON file
     * @param filePath Path to the JSON file (relative to project root or absolute)
     */
    constructor(filePath: string = "../puzzles.json") {
        const resolvedPath = path.resolve(__dirname, filePath);

        if (!fs.existsSync(resolvedPath)) {
            throw new Error(`Puzzle file not found: ${resolvedPath}`);
        }

        const fileContent = fs.readFileSync(resolvedPath, "utf-8");
        const data: PuzzleCollection = JSON.parse(fileContent);

        this.puzzles = data.puzzles;
        this.validatePuzzles();
    }

    /**
     * Validates that all loaded puzzles have valid 9x9 grids
     */
    private validatePuzzles(): void {
        this.puzzles.forEach((puzzle, index) => {
            if (!puzzle.grid || puzzle.grid.length !== GRID_SIZE) {
                throw new Error(`Puzzle "${puzzle.name}" (index ${index}) must have exactly 9 rows`);
            }

            puzzle.grid.forEach((row, rowIndex) => {
                if (row.length !== GRID_SIZE) {
                    throw new Error(
                        `Puzzle "${puzzle.name}" row ${rowIndex} must have exactly 9 columns`
                    );
                }

                row.forEach((cell, colIndex) => {
                    if (!Number.isInteger(cell) || cell < EMPTY_CELL || cell > MAX_DIGIT) {
                        throw new Error(
                            `Puzzle "${puzzle.name}" has invalid value at [${rowIndex}][${colIndex}]: ${cell}`
                        );
                    }
                });
            });
        });
    }

    /**
     * Gets all loaded puzzles
     */
    public getAllPuzzles(): Puzzle[] {
        return [...this.puzzles];
    }

    /**
     * Gets a puzzle by its name
     * @param name The name of the puzzle to retrieve
     * @returns The puzzle if found, undefined otherwise
     */
    public getPuzzleByName(name: string): Puzzle | undefined {
        return this.puzzles.find(p => p.name === name);
    }

    /**
     * Gets puzzles by difficulty level
     * @param difficulty The difficulty level (e.g., "easy", "medium", "hard")
     */
    public getPuzzlesByDifficulty(difficulty: string): Puzzle[] {
        return this.puzzles.filter(p => p.difficulty.toLowerCase() === difficulty.toLowerCase());
    }

    /**
     * Gets a puzzle by index
     * @param index Zero-based index of the puzzle
     */
    public getPuzzleByIndex(index: number): Puzzle | undefined {
        return this.puzzles[index];
    }

    /**
     * Gets the total number of loaded puzzles
     */
    public getPuzzleCount(): number {
        return this.puzzles.length;
    }

    /**
     * Lists all puzzle names
     */
    public listPuzzleNames(): string[] {
        return this.puzzles.map(p => p.name);
    }
}

File: demo-apps/demoapp002-python-pytest/app_src/puzzle_loader.py
from __future__ import annotations

import json
from copy import deepcopy
from dataclasses import dataclass
from pathlib import Path

from .constants import EMPTY_CELL, GRID_SIZE, MAX_DIGIT


@dataclass(frozen=True)
class Puzzle:
    name: str
    difficulty: str
    description: str
    grid: list[list[int]]


class PuzzleLoader:
    def __init__(self, file_path: str = "../puzzles.json"):
        path = Path(file_path)
        if not path.is_absolute():
            path = Path(__file__).resolve().parent / file_path
        path = path.resolve()

        if not path.exists():
            raise FileNotFoundError(f"Puzzle file not found: {path}")

        data = json.loads(path.read_text(encoding="utf-8"))
        self._puzzles = [
            Puzzle(
                name=item["name"],
                difficulty=item["difficulty"],
                description=item["description"],
                grid=deepcopy(item["grid"]),
            )
            for item in data["puzzles"]
        ]
        self._validate_puzzles()

    def get_all_puzzles(self) -> list[Puzzle]:
        return deepcopy(self._puzzles)

    def get_puzzle_by_name(self, name: str) -> Puzzle | None:
        return next((puzzle for puzzle in self._puzzles if puzzle.name == name), None)

    def get_puzzles_by_difficulty(self, difficulty: str) -> list[Puzzle]:
        return [puzzle for puzzle in self._puzzles if puzzle.difficulty.lower() == difficulty.lower()]

    def get_puzzle_by_index(self, index: int) -> Puzzle | None:
        return self._puzzles[index] if 0 <= index < len(self._puzzles) else None

    def get_puzzle_count(self) -> int:
        return len(self._puzzles)

    def list_puzzle_names(self) -> list[str]:
        return [puzzle.name for puzzle in self._puzzles]

    def _validate_puzzles(self) -> None:
        for index, puzzle in enumerate(self._puzzles):
            if len(puzzle.grid) != GRID_SIZE:
                raise ValueError(f'Puzzle "{puzzle.name}" (index {index}) must have exactly 9 rows')
            for row_index, row in enumerate(puzzle.grid):
                if len(row) != GRID_SIZE:
                    raise ValueError(f'Puzzle "{puzzle.name}" row {row_index} must have exactly 9 columns')
                for col_index, cell in enumerate(row):
                    if not isinstance(cell, int) or cell < EMPTY_CELL or cell > MAX_DIGIT:
                        raise ValueError(
                            f'Puzzle "{puzzle.name}" has invalid value at [{row_index}][{col_index}]: {cell}'
                        )
